Ashlar module

jmuhlich opened the pull request:
Temporary PR for code review purposes.

RobJY committed:
working on ashlar nextflow module outside nf-core template.  Input and output working.

RobJY committed:
added default output filename.

RobJY committed:
changed output parameter to match ashlar param name and changed output path to allow for both output file types

RobJY committed:
added optional input parameter channel

RobJY committed:
added most input paramters. working code

RobJY committed:
clean-up

RobJY committed:
added output-channels parameter

RobJY committed:
added support for multiple input files

RobJY committed:
added ffp input parameter

RobJY committed:
bug fix: should not have had multiple channels

RobJY committed:
most of meta.yml filled out. still need the case for *.tif output

RobJY committed:
ah, also need to add defaults and optional args to meta.yml

RobJY committed:
minimal set of input parameters added to meta.yml for testing; need to set defaults for them

RobJY committed:
runs simple test with hard-coded paths

RobJY committed:
simple test works with input parameter and no hard-coded paths

RobJY committed:
added ashlar main module file

RobJY committed:
added test image path to test workflow so you can run with default test command

RobJY committed:
added ext.args

RobJY committed:
added meta map

RobJY committed:
passed test with wildcarded input files

RobJY committed:
added test config file

RobJY committed:
added functionality to handle optional arguments

RobJY committed:
added meta map args

RobJY committed:
supposed to output meta map if you input one

RobJY committed:
clean-up

RobJY committed:
changed test input to a channel

RobJY committed:
not running, but close I think

RobJY committed:
added some missing files

RobJY committed:
basic samplesheet working

RobJY committed:
clean-up post testing

RobJY committed:
enabled multiple input images

RobJY committed:
bug fix: sheet meta arg not an it var

RobJY committed:
clean-up

RobJY committed:
added samplesheet validation

RobJY committed:
samplesheet output to subdirs with sample name; needs clean-up

RobJY committed:
cleaned up meta.yml and passed test

RobJY committed:
clean-up; passed tests

RobJY committed:
moved publishDir to test config as suggested by lint

RobJY committed:
removed initOptions; lint says deprecated; tests pass

RobJY committed:
multi-file row of test_1 failing, but others work

RobJY committed:
passes 3 tests with stable md5 sums

RobJY committed:
clean-up and changed back to uuid block zero out instead of while loop; added exception check

RobJY committed:
removed test publishDir for sample sheets

RobJY committed:
clean-up

RobJY committed:
added Dockerfile

RobJY committed:
clean-up

jmuhlich commented on .nf-core.yml:
```diff
@@ -8,3 +8,4 @@ bump-versions:
   rseqc/inferexperiment: False
   rseqc/innerdistance: False
   sortmerna: False
+org_path: nf-core
```
What is this new line?

RobJY replied:
Thank you for pointing this out.  I don't think it should be included in the commit.  That line is created when running `nf-core modules create-test-yml ashlar`, so you don't have to enter the org_path each time.  This may be different for other users though, so that's why it isn't in the master branch.  I've removed it now and the tests run fine as expected.

jmuhlich commented on modules/nf-core/ashlar/Dockerfile:
```diff
@@ -0,0 +1,27 @@
+FROM ubuntu:22.04
```
I don't think we need the Dockerfile in here?

RobJY replied:
In the [Software Requirements section](https://nf-co.re/developers/modules#software-requirements) of the [DSL2 Modules documentation](https://nf-co.re/developers/modules), point 5 says:
> If the software is not available on Bioconda a Dockerfile MUST be provided within the module directory. We will use GitHub Actions to auto-build the containers on the GitHub Packages registry.

I have a question up on their slack about what the `container` call should be, but they haven't responded yet.

jmuhlich commented on modules/nf-core/ashlar/ashlar.nf:
```diff
@@ -0,0 +1,92 @@
+process wc {
```
This module was using the old approach and can be removed, right?

RobJY replied:
You're right.  I've removed it now.

jmuhlich commented on modules/nf-core/ashlar/main.nf:
```diff
@@ -0,0 +1,40 @@
+    label 'process_single'
+
+    conda (params.enable_conda ? "bioconda::ashlar=1.17.0" : null)
+    /* TODO nf-core: we'll provide a dockerfile as described in 5) under Software Requirements
```
Since Ashlar (and most of our own modules) are written in Python, it actually wouldn't be a big lift to make them available via bioconda. I guess we need to consider whether having duplicate containers aside from our own "official" ones is OK, or whether we just want to delegate to bioconda as our official containers anyway.

RobJY replied:
Yeah, adding to bioconda would streamline this process, but as I mention in the question above, there's supposed to be a way for us to provide a Dockerfile that they then build and maintain a container for.  Hopefully they'll respond to my question about this soon.

RobJY replied:
I still haven't gotten any reply to my question about this on the nf-core Slack modules channel.  I went ahead and tried using our Docker Hub image and that seems to work fine.  It passed both the lint and functional tests.  I've posted another question on Slack asking if this is acceptable or if they'd prefer we provide a Dockerfile.

RobJY replied:
I got a quick response this time:

> So, nf-core modules usually use the Biocontainers that are automatically produced by the Bioconda tooling. That helps us be sure of consistency, and you also get the singularity container as a bonus (if you look at a couple of modules you’ll see what I mean).
Yours should be present at https://quay.io/repository/biocontainers/ashlar. I can’t see it, which probably means it’s not been made public (which unfortunately doesn’t happen by default). If you go to https://gitter.im/biocontainers/Lobby and ask nicely, someone (probably Bjoern) will make it public for you, and you can use it (and the singularity container) here. You also won’t have to maintain your own version of the container (unless you have other reasons for doing so).

So it does look like ashlar is there.  I found it at:
https://bioconda.github.io/recipes/ashlar/README.html?highlight=ashlar#package-package%20&#x27;ashlar&#x27;
but when I try the container pull command on the page I get:
```
(ashlar-recipe) pollen@system76-pc:/tmp$ docker pull quay.io/biocontainers/ashlar:latest
Error response from daemon: unauthorized: access to the requested resource is not authorized
```
Shall I go ahead and request that it be made public as suggested?

jmuhlich commented on modules/nf-core/ashlar/meta.yml:
```diff
@@ -0,0 +1,44 @@
+      description: "Alignment by Simultaneous Harmonization of Layer/Adjacency Registration"
+      homepage: "https://labsyspharm.github.io/ashlar/"
+      documentation: "https://labsyspharm.github.io/ashlar/"
+      doi: "10.1101/2021.04.20.440625"
```
Final non-preprint DOI is `10.1093/bioinformatics/btac544`

RobJY replied:
updated that.  Thanks!

jmuhlich commented on modules/nf-core/ashlar/meta.yml:
```diff
@@ -0,0 +1,44 @@
+  - file_in:
+      type: path
+      description: multidimentional image data in formats from most commercial microscopes
+      pattern: "*.ome.tiff"
```
Does this need to be exhaustive? There are _many_ possible extensions here. In practice no microscopes actually generate ome-tiffs -- I just converted our exemplars to that format for the sake of using an open format.

RobJY replied:
Hmm, it might need to be exhaustive.  I'll look into this more.

RobJY replied:
Yes, I think it's supposed to be exhaustive, but there doesn't seem to be a check when running the tests.  They state that the tests check that the input/output information between `main.nf` and `meta.yml` is consistent, but I messed around with the input file suffixes and didn't get any errors when testing.  I think we should add something like what other projects have in their `meta.yml` files like:
```
      pattern: "*.{fa,fasta,fna,fa.gz,fasta.gz,fna.gz,gbk,gbk.gz,embl,embl.gz}"
```
I've put a start of this in the last commit and it passed tests.  Can you provide a list of all or most of the possible input file extensions?  I don't think it will stop anything from working if we miss some and add them at a later date.  I should probably test this a bit more though.  Could you provide one or two small files of different types to test with?

Reading over the Ashlar documentation I realize I haven't tested for this case:
> If PATH ends in just .tif and includes {cycle} and {channel} placeholders, a series of single-channel plain TIFF files will be written.

Can you provide a good test .tif file for this?

jmuhlich replied:
PATH meaning the output path, not input (as PATH is the metavar for the -o help section). E.g. if you write `ashlar inputs... -o out_{cycle}_{channel}.tif` you will get `out_0_0.tif`, `out_0_1.tif`, etc.

jmuhlich replied:
```pattern: "*.{ome.tiff,ome.tif,rcpnl,btf,nd2,tiff,tif,czi}"```
We also support `*.{xdce,nd,scan,htd}` but these are trickier because those extensions just define "index" files and all the actual image tiles come in separate individual .tif files, so the whole containing directory needs to be staged not just the index file. You can look at the existing mcmicro code to see how we do this, but I might need to explain it because it's baroque :rofl: .

RobJY replied:
> PATH meaning the output path, not input (as PATH is the metavar for the -o help section). E.g. if you write `ashlar inputs... -o out_{cycle}_{channel}.tif` you will get `out_0_0.tif`, `out_0_1.tif`, etc.

Oh, I see.  Thanks for explaining that.  I'll try adding a test for this.

jmuhlich replied:
See my other comments - I don't think this feature needs to be supported in this nf-core module

RobJY replied:
Ok, sounds good.

RobJY replied:
updated, tested and pushed.

jmuhlich commented on modules/nf-core/ashlar/meta.yml:
```diff
@@ -0,0 +1,44 @@
+  - file_out:
+      type: file
+      description: pyramidal OME-TIFF
+      pattern: "*.tif"
```
This should be `*.ome.tif`.

RobJY replied:
I set it as `*.tif` because the output can be `.tif` if there's a non-ome.tif input.  That said I realize that the input should be changed to `*.tif` because of this.  I'll make that change.
Thank you for pointing this out.

RobJY replied:
tests passed as expected with this change, so I've pushed the commit.

jmuhlich replied:
As I commented above, the input name does not affect the output name. The output are only non-.ome.tif if the user puts `{cycle}` and/or `{channel}` in the _output_ file path (-o). I am tempted to just not support this option in the nf-core module. If someone wants to split their all-in-one ome-tiff into single channels or whatever, they can use a separate module to post-process it. mcmicro will always use the all-in-one ome-tiff so we don't need the single-channel tif output anyway. I don't think we need to expose the `-o` option to module users at all, actually.

jmuhlich replied:
Therefore the file_out pattern should be `*.ome.tif`.

RobJY replied:
Sounds good. updated, tested and pushed

jmuhlich commented on tests/modules/nf-core/ashlar/input_sheet.csv:
```diff
@@ -0,0 +1,3 @@
+sample,file_list,args
```
Do we still need to provide this file for the module?

RobJY replied:
You're correct, we don't need that now.  Thanks!
I've removed it.

jmuhlich commented on tests/modules/nf-core/ashlar/main.nf:
```diff
@@ -0,0 +1,61 @@
+
+include { ASHLAR } from '../../../../../modules/modules/nf-core/ashlar/main.nf'
+// we zero out the UUID of output tiff images with ZERO_UUID so we get a consistent md5sum
+include { ZERO_UUID } from './zero_uuid.nf'
```
I think you forgot to `git add zero_uuid.nf` -- I don't see it here?

RobJY replied:
You're right, I forgot to add it.  Thanks!
It's in there now.  Please take a look when you've got a chance.

jmuhlich commented on tests/modules/nf-core/ashlar/main.nf:
```diff
@@ -0,0 +1,61 @@
+// we zero out the UUID of output tiff images with ZERO_UUID so we get a consistent md5sum
+include { ZERO_UUID } from './zero_uuid.nf'
+
+def TEST_IMG = "/home/pollen/HITS/nextflow/mcmicro/exemplar-001/raw/exemplar-001-cycle-0{6,7}.ome.tiff"
```
Absolute paths won't work here. How do other module tests handle paths?

RobJY replied:
right, we'll need to add our test images to their test file directory in this repo, but I need to reach out to them to see if I should do that as part of this PR or before.  I was going to contact nf-core about it after we settled on smaller test images, so I could tell them the sizes.
I should have mentioned that in the code though.

RobJY replied:
I just successfully tested using the new images you posted.  Will update with relative paths when they are merged.

jmuhlich reviewed (commented):
One thing I didn't want to leave in a per-line comment since the diff is too big to render by default: Did `tests/config/pytest_modules.yml` get auto-reformatted by your editor? Looks like there are many whitespace changes that we don't actually want to introduce.

I still need to look over the samplesheet related stuff, but maybe those are actually out of scope here?

RobJY committed:
updated to correct DOI

RobJY committed:
added zero_uuid.nf

RobJY committed:
removed test sample sheet

RobJY committed:
changed input type to *.tif to match output type

RobJY committed:
removed old test ashlar nextflow module

RobJY committed:
Update .nf-core.yml

removed added line from mistaken commit.

RobJY commented:
> One thing I didn't want to leave in a per-line comment since the diff is too big to render by default: Did tests/config/pytest_modules.yml get auto-reformatted by your editor? Looks like there are many whitespace changes that we don't actually want to introduce.

Hmmm, from point 6 under [New module workflow](https://nf-co.re/developers/modules#new-module-workflow) in the [DSL2 Modules documentation](https://nf-co.re/developers/modules), it looks like that was automatically edited when I ran the command to create the new module.  I don't remember editing it manually, but you're right, it looks like something an editor would do.  I'll change the formatting back to the way it was.

update:
updated formatting, passed tests and pushed code

> I still need to look over the samplesheet related stuff, but maybe those are actually out of scope here?

Yeah, we're not submitting any examples of using the sample sheets and there's an issue with zeroing out the UUID for one of the test cases, so no need to look at it for this PR.

RobJY committed:
restored formatting in pytest_modules.yml

RobJY committed:
added more accurate input pattern

RobJY committed:
works with new test images

jmuhlich commented on tests/modules/nf-core/ashlar/zero_uuid.nf:
```diff
@@ -0,0 +1,32 @@
+
+    script:
+    """
+    TARG_PATHS_ARRAY=(\$(find ../.. -name "ashlar_output.ome.tif" -printf "%k %p\n" | sort -n | awk -F ' ' '{print \$2}'))
```
Even though I think we want to take a totally different approach here, I want to point out a general Nextflow thing here: This relative path access is not generally OK! You can only rely on what you were given in the `input:` file args -- traversing the filesystem is not going to work when the files were staged from e.g. cloud storage and that other file isn't present in the local work directory.

RobJY replied:
Oh, sure, thanks for pointing that out.  I was thinking that this directory structure would always be present, but that makes sense.

As you say, working from hard-coded offsets makes much more sense anyway.

jmuhlich commented on tests/modules/nf-core/ashlar/zero_uuid.nf:
```diff
@@ -0,0 +1,32 @@
+process ZERO_UUID {
+
+    input:
+    each path(file_in)
```
I propose this takes a file path _and_ the numeric offset of the UUID, which we will pre-calculate for each test output file and hard-code in the test script calls to ZERO_UUID. That simplifies this code dramatically. We could even write this purely in Groovy since it's just some simple file operations, but that optimization can be done some other time.

RobJY replied:
Ok, sure, that makes sense.  sounds good to me.

RobJY replied:
I just pushed this refactor.  Thanks so much for the suggestion.  It's so much cleaner.

RobJY replied:
I just pushed a simplified version of zero_uuid.nf that gets the input file path in groovy so I don't have to parse it out in bash.

jmuhlich commented on modules/nf-core/ashlar/main.nf:
```diff
@@ -0,0 +1,33 @@
+
+    script:
+    def args_conf = task.ext.args ?: ''
+    def args_meta = meta.args ?: ''
```
I just realized putting args in meta doesn't make sense, since the sample sheet will be one row per cycle, and the args need to be per ashlar call i.e. per experiment (many cycles).

RobJY replied:
Don't the meta args apply to the ashlar call?
In the existing (non-sample sheet) code in main.nf (right after the commented code above) we have:
```
    ashlar \\
        $file_in \\
        $args_conf \\
        $args_meta
```
Then in the sample sheet version (in the ashlar_sheet branch) we have additional args for each row.  So maybe we don't need the additional args in the sample sheet?
Am I misunderstanding this?

RobJY committed:
added test.yml

jmuhlich commented on tests/modules/nf-core/ashlar/main.nf:
```diff
@@ -0,0 +1,78 @@
+
+workflow test_ashlar_one_file {
+
+    input_list =  [ [ [ id:'test_1', args: '--flip-y' ],
```
We can't use --flip-y and --flip-x when the input image doesn't need it. The output will be nonsensical.

RobJY replied:
Ok, sounds good.  I was just looking for something simple to test with.
I can just do a test with no args and the 3 cycif tonsil images, but I'd also like to have a test of the args (I see above that we'll need to change how we handle them though).
Would using --filter-sigma or --tile-size make sense as a test parameter?  If so, what's a reasonable value to use?

jmuhlich replied:
Adding a --filter-sigma value should change the output image pixels very slightly, but maybe this synthetically generated dataset is "too perfect" for it to have any effect. --tile-size will change the overall structure of the TIFF file regardless so maybe that's a better test.

RobJY replied:
Sounds good.  Thanks!
I've added a tile-size test.

RobJY committed:
removed unsupported output test and updated input and output patterns in meta; tests pass

RobJY committed:
refactored zero_uuid.nf to use input parameter offset; passed test

RobJY committed:
added tile size test as discussed in PR

RobJY committed:
bug fix: off by one error for second test offset

RobJY committed:
clean-up

RobJY committed:
simplified zero_uuid.nf

RobJY committed:
Merge branch 'nf-core:master' into ashlar

RobJY committed:
changed to built-in test files

+ use built-in test files
+ changed input parameter to path from val
+ removed input parameter from nextflow.confg

RobJY committed:
removed conda command so runs locally

Sorry, didn't mean to push the last commit.
This one has the changes and runs fine locally.  The previous commit fails because of 'conda "bioconda::ashlar=1.17.0"' in main.nf.  Seems that it fails to build the local conda environment.

RobJY committed:
added commented code that causes break

RobJY committed:
works in linux with conda so uncommented that section

RobJY committed:
added containers and tested all 3 ways of running; all working

RobJY committed:
ran prettier

RobJY committed:
updated docker container URL

RobJY commented:
I was able to fix the prettier error we were getting easily, but I'm having more trouble with the issue that popped up since they changed the default behavior for the containers.

I'm able to run `nf-core modules create-test-yml ashlar` successfully locally with the line `labsyspharm/ashlar:1.17.0` in main.nf, but this fails for the CI tests on GitHub with an error `Unable to connect to container URL`.  Interestingly I can't run locally if I put the hash in the above line even though it hasn't changed at https://quay.io/repository/biocontainers/ashlar?tab=tags&tag=latest.

Looks like with this new container URL change it should be easier to use containers from the Docker registry, but I haven't had any luck yet with getting that to work.

I'll keep at it, but please let me know if you see anything I'm doing wrong.  Thanks!

RobJY committed:
testing container URL with hash

RobJY committed:
switched container to docker.io

RobJY commented:
Got the docker.io URL to work locally with `nf-core modules create-test-yml ashlar` by removing the hash.  Still getting an error from the CI, but I see it's a 403.  Why would we be getting `forbidden` from docker.io for our container when run from the CI?  Looking into it.

RobJY committed:
fixed meta.yaml lint errors

RobJY commented:
got the lint errors for meta.yml fixed.  Still working on that 403 from docker.io.

RobJY committed:
trying CI again with both container hashes

RobJY committed:
trying with default container registry

RobJY committed:
bug fix: removed extra char output in version.yml

RobJY committed:
changed container URL syntax and it works locally; testing with CI

RobJY committed:
added conda again to test with CI

RobJY commented:
Hmm, I don't understand why we're failing this last test.  It runs fine locally and the error message from the CI looks suspicious, so I've posted a [question on their Slack channel](https://nfcore.slack.com/archives/CE6PELWR4/p1684357273172049).

RobJY committed:
trying container URL without quay.io again

RobJY committed:
Merge branch 'master' into ashlar

RobJY committed:
testing URL with quay.io again

RobJY commented:
Hmm, we passed the CI container tests with code that passes the local lint tests, but fails the local `nf-core modules create-test-yml ashlar` test with a 'no such image' error.

I'll keep working with nf-core on Slack to try to track this down.

RobJY commented:
`nf-core modules create-test-yml ashlar` is now working.  I just tried it out before asking them for updates on the Slack channel and it works.  I didn't change anything and it doesn't look like they did, so I'm a bit confused.  I guess something could have changed at biocontainers that fixed it, but it's still strange.  It's typically the first test I run, so I'll notice quickly if it reappears.

I pulled updates of Nextflow and nf-validation and now I'm getting a different error message when I try to run nf-validation.  I'll look into this more and post a question on Slack if needed.

RobJY committed:
added singularity; passes local tests

RobJY committed:
Merge branch 'master' into ashlar

RobJY commented:
I realized I should merge in the changes from origin and make sure nothing breaks, but all looks good.

RobJY committed:
Merge branch 'master' into ashlar

RobJY commented:
I updated from origin/main again and we still pass all tests.
Please let me know if you'd like any changes before I make a PR with nf-core.  Thanks!

RobJY committed:
cleaned-up unused files

RobJY committed:
removed meta args

RobJY commented:
Hmm, we failed CI tests for docker because of the output test.yml file which is automatically generated.
I just pushed the commit that removes the meta args and we'll see if that's still an issue.

RobJY commented:
Ok, all green from the CI after the update to remove the meta args.

jmuhlich commented on modules/nf-core/ashlar/main.nf:
```diff
@@ -0,0 +1,33 @@
+    task.ext.when == null || task.ext.when
+
+    script:
+    def args_conf = task.ext.args ?: ''
```
This could just be called `args` now like other simple modules since we dropped the other args variable.

RobJY replied:
agreed. I've made the change and it passes the tests.

jmuhlich commented on modules/nf-core/ashlar/main.nf:
```diff
@@ -0,0 +1,33 @@
+    tuple val(meta), path(file_in)
+
+    output:
+    tuple val(meta), path("*.tif")      ,   emit: ashlar_tif
```
Maybe this should just be `emit: tif`. Reviewing a bunch of other modules, I almost never see the tool name used in output names. Also the pattern should really be `*.ome.tif`.

RobJY replied:
Yes, you're right.  I've made the change and it passes the tests.

jmuhlich commented on modules/nf-core/ashlar/main.nf:
```diff
@@ -0,0 +1,33 @@
+
+    cat <<-END_VERSIONS > versions.yml
+    "${task.process}":
+        ashlar: \$(echo \$(ashlar --version 2>&1) | sed 's/^.*ashlar //; s/Using.*\$//' )
```
Can simplify to:
```bash
\$(ashlar --version 2>&1 | sed 's/^.*ashlar //; s/Using.*\$//' )
```

RobJY replied:
Cool, thanks!  I made the change and it passes the tests.

jmuhlich commented on modules/nf-core/ashlar/meta.yml:
```diff
@@ -0,0 +1,45 @@
+      homepage: "https://labsyspharm.github.io/ashlar/"
+      documentation: "https://labsyspharm.github.io/ashlar/"
+      doi: "10.1093/bioinformatics/btac544"
+      licence: "['MIT']"
```
This looks over-quoted? Probably should be `["MIT"]`

RobJY replied:
You're right, That looks much better and passes the tests.
I had copied what I have from some other modules, but I see examples of what you have after looking at more of them.

jmuhlich commented on modules/nf-core/ashlar/meta.yml:
```diff
@@ -0,0 +1,45 @@
+      licence: "['MIT']"
+
+input:
+  #
```
Spurious empty comment line

RobJY replied:
agreed, thanks!

jmuhlich commented on modules/nf-core/ashlar/meta.yml:
```diff
@@ -0,0 +1,45 @@
+      type: map
+      description: |
+        Groovy Map containing sample information
+        e.g. [ id:'test', : args: '--flip-x' ]
```
Delete `args` from example.

RobJY replied:
Oops, yes, I forgot we had an example in meta.yml.  Thanks for catching it!

jmuhlich commented on modules/nf-core/ashlar/meta.yml:
```diff
@@ -0,0 +1,45 @@
+        e.g. [ id:'test', : args: '--flip-x' ]
+  - file_in:
+      type: file
+      description: multidimentional image data in formats from most commercial microscopes
```
Multidimensional is probably the wrong term. Let's use "Overlapping tile image data...".

RobJY replied:
Right, that makes sense.  I've made the change.  Thanks!

jmuhlich commented on modules/nf-core/ashlar/meta.yml:
```diff
@@ -0,0 +1,45 @@
+      description: multidimentional image data in formats from most commercial microscopes
+      pattern: "*.{ome.tiff,ome.tif,rcpnl,btf,nd2,tiff,tif,czi}"
+output:
+  #
```
Spurious empty comment

RobJY replied:
agreed. Thanks!

jmuhlich commented on modules/nf-core/ashlar/meta.yml:
```diff
@@ -0,0 +1,45 @@
+      type: map
+      description: |
+        Groovy Map containing sample information
+        e.g. [ id:'test', : args: '--flip-x' ]
```
Delete `args` from example

RobJY replied:
oops, yes, I forgot this template was also in meta.yml and would need to be updated.  thank you for catching it!

jmuhlich commented on modules/nf-core/ashlar/meta.yml:
```diff
@@ -0,0 +1,45 @@
+      description: |
+        Groovy Map containing sample information
+        e.g. [ id:'test', : args: '--flip-x' ]
+  - file_out:
```
Shouldn't this match the actual output arg? It was `ashlar_tif` but if you accept my change above it will become just `tif`.

RobJY replied:
Ah, yes, absolutely.  Thanks for catching it!  I should have realized that.

jmuhlich reviewed (commented):
A bunch of small tweaks. almost there!

RobJY committed:
removed unnecessary Channel.toList from test main

RobJY committed:
restored tile test using withName technique for param

RobJY committed:
simplified input arg name

RobJY committed:
updated module output path and emit

RobJY committed:
updated module usage command

RobJY committed:
updated licence in meta

RobJY committed:
cleaned-up meta.yml

RobJY committed:
Merge branch 'master' into ashlar

RobJY commented:
Our PR with nf-core showed this branch as being out of sync by a few commits, so I added them to make sure that nothing in them broke our code.  We passed all local and CI tests.

RobJY committed:
Update modules/nf-core/ashlar/main.nf

Co-authored-by: Simon Pearce <24893913+SPPearce@users.noreply.github.com>

RobJY committed:
Update modules/nf-core/ashlar/main.nf

Co-authored-by: Simon Pearce <24893913+SPPearce@users.noreply.github.com>

RobJY committed:
Update modules/nf-core/ashlar/main.nf

Co-authored-by: Simon Pearce <24893913+SPPearce@users.noreply.github.com>

RobJY committed:
changed dfp and ffp to optional input channels and tested with no input files

RobJY committed:
added dfp ffp test with hard-coded files

SPPearce committed:
Merge branch 'master' into ashlar

RobJY committed:
added validation for dfp and ffp

RobJY committed:
changed validation for dfp/ffp and added test

RobJY commented:
I pushed an update that changes the dfp/ffp validation and adds a test.
I got the following error when I tried to use the same dfp file multiple times to make an input list:
```
CRITICAL Error running test workflow (exit code 1)                                                             
         N E X T F L O W  ~  version 23.04.1                                                                   
         Launching `./tests/modules/nf-core/ashlar/main.nf` [peaceful_ptolemy] DSL2 - revision: 648ea093c8     
         ERROR ~ Error executing process > 'test_ashlar_1_file_dfp_ffp:ASHLAR'                                 
                                                                                                               
         Caused by:                                                                                            
           Process `test_ashlar_1_file_dfp_ffp:ASHLAR` input file name collision -- There are multiple input   
         files for each of the following file names: cycif-tonsil-small-dfp.tif, cycif-tonsil-small-ffp.tif    
```
Because of this I changed the validation to accept 0, 1 or n where n == input file list length.
Please let me know if this is acceptable or if you'd rather it was done differently.  Thanks!

RobJY commented:
We're failing CI because I'm just testing with local copies of the dfp/ffp files until they're available in the test-data directory.
The code is passing both the lint and pytests on my machine.

RobJY committed:
Merge branch 'master' into ashlar

RobJY commented:
I've pushed an update to the ashlar_validation branch that includes the nf-validation plug-in help and parameter validation functions applied to the ashlar branch code.  You can test it with the commands:
`nextflow run -entry test_ashlar_1_file -plugins nf-validation tests/modules/nf-core/ashlar/main.nf -c tests/config/nextflow.config --help`
`nextflow run -entry test_ashlar_1_file -plugins nf-validation tests/modules/nf-core/ashlar/main.nf -c tests/config/nextflow.config --validate_params`
The above commands assume you're in the base modules directory.
The --validate_params command will not output anything when everything is configured correctly.

Should I pull this into the ashlar branch or is it unnecessary?
I know we'll want it for the later workflows, but I wanted to get it working for the basic ashlar test workflow before getting it working for the samplesheet test code.  It's been more simple than going straight to the samplesheet code and it's helped to flush out some errors I had in our config files.

RobJY commented:
I've updated the ashlar_sheet branch.  It contains working code for using sample sheet input.  I pulled in the validation updates and finally got the md5sum issue fixed.
The next step in this branch is to get Channel.fromSamplesheet from the nf-validation plug-in working.  This should clean up the code quite a bit because we'll be able to remove the existing code I lifted from the rnaseq repo for doing this.
I also need to remove the args column from the sample sheet and add columns for dfp and ffp files.

**Note:**
To run the test with: `nf-core modules create-test-yml ashlar`
you'll need to input the following when asked for the 'Test command' instead of accepting the default:
`nextflow run -entry test_ashlar_sheet -plugins nf-validation tests/modules/nf-core/ashlar/main.nf -c tests/config/nextflow.config --input <enter path here>/modules/tests/modules/nf-core/ashlar/test_sheet_url.csv`
The above just adds the `--input` flag with path to the sample sheet.

RobJY committed:
Merge branch 'master' into ashlar

RobJY committed:
now pulling dfp/ffp from test_data repo

RobJY commented:
I pushed updated code that uses the new test data and we're passing all CI and local tests.

Sorry, I didn't realize I needed to make the change to the nextflow.config file to add the new test files.  I thought that was pushed out by the CI.  I asked on Slack and they set me straight.

Please let me know if this looks good to go.
When it is I'll update our PR with nf-core and see if they're ready to merge it or would like more changes.  Thanks!

RobJY committed:
Merge branch 'master' into ashlar

RobJY committed:
Merge branch 'master' into ashlar

RobJY committed:
Merge branch 'master' into ashlar

RobJY committed:
test_data formatting change

RobJY committed:
Merge branch 'master' into ashlar

RobJY committed:
Merge branch 'master' into ashlar

RobJY committed:
Merge branch 'master' into ashlar

RobJY committed:
Update modules/nf-core/ashlar/main.nf

moved dfp/ffp validation into script block

Co-authored-by: Adam Talbot <12817534+adamrtalbot@users.noreply.github.com>

RobJY committed:
corrected dfp/ffp error message

RobJY committed:
Update modules/nf-core/ashlar/meta.yml

corrected input name

Co-authored-by: Adam Talbot <12817534+adamrtalbot@users.noreply.github.com>

RobJY committed:
Update modules/nf-core/ashlar/meta.yml

made file description more informative

Co-authored-by: Adam Talbot <12817534+adamrtalbot@users.noreply.github.com>

RobJY committed:
finished input name change

RobJY committed:
Apply suggestions from code review

nice, I didn't realize it could be referenced like that. Thanks!

Co-authored-by: Adam Talbot <12817534+adamrtalbot@users.noreply.github.com>

RobJY committed:
Update tests/modules/nf-core/ashlar/main.nf

Ah, sure, that makes sense.  I was trying to abstract it away, but I see how it's nice to see it all at once. Thanks!

Co-authored-by: Adam Talbot <12817534+adamrtalbot@users.noreply.github.com>

RobJY committed:
broken change for discussion

RobJY committed:
Update tests/modules/nf-core/ashlar/main.nf

Oh! sorry, I see. I missed the `tuple` before `val(meta), path(tif)` when I was testing the change.  I should have realized that!

Co-authored-by: Adam Talbot <12817534+adamrtalbot@users.noreply.github.com>

RobJY committed:
fixed zero_uuid output bug

RobJY committed:
Merge branch 'master' into ashlar

RobJY committed:
Merge branch 'master' into ashlar

RobJY committed:
Merge branch 'master' into ashlar

RobJY committed:
Merge branch 'master' into ashlar

RobJY committed:
Merge branch 'master' into ashlar

SPPearce committed:
Use sed to remove UUID inside main module

SPPearce committed:
Fixed sed

SPPearce committed:
Update modules/nf-core/ashlar/main.nf

Co-authored-by: Adam Talbot <12817534+adamrtalbot@users.noreply.github.com>

RobJY committed:
Merge branch 'master' into ashlar

RobJY committed:
proposed change to sed command

RobJY commented:
Hmm, I spoke too soon.  It passed my quick local test, but is failing the CI tests.  Looking into it now.

RobJY committed:
adding modified test.yml file

RobJY committed:
reverted to in-place sed

RobJY committed:
Merge branch 'master' into ashlar
